src/mito2/src/manifest/storage.rs

```diff
@@ -405,9 +411,23 @@ impl ManifestObjectStore {
         let path = self.checkpoint_file_path(version);
         // Due to backward compatibility, it is possible that the user's checkpoint not compressed,
         // so if we don't find file by compressed type. fall back to checkpoint not compressed find again.
-        let checkpoint_data =
-            match self.object_store.read(&path).await {
-                Ok(checkpoint) => {
+
+        // get the stored_checksum, which is in the checkpoint metadata.
+        let metadata_path = self.last_checkpoint_path();
+        let metadata_bytes = match self.object_store.read(&metadata_path).await {
+            Ok(metadata_bytes) => metadata_bytes,
+            Err(e) if e.kind() == ErrorKind::NotFound => return Ok(None), // No metadata found
+            Err(e) => return Err(e).context(OpenDalSnafu),                // Error reading metadata
+        };
+        let metadata = CheckpointMetadata::decode(&metadata_bytes)?;
+
+        let checkpoint_data = match self.object_store.read(&path).await {
+            Ok(checkpoint) => {
+                let mut hasher = Hasher::new();
+                hasher.update(&checkpoint);
+                let calculated_checksum = hasher.finalize();
+
+                if Some(format!("{}", calculated_checksum)) == metadata.checksum {
```

Comment:
What if `metadata.checksum` is `None`? We should not fail in this case

Reply:
Thanks for your valuable suggestions!  I'll work on them.